L74-java-channel-creds.md

```diff
@@ -0,0 +1,571 @@
+   * returned, even though only one may be necessary.
+   *
+   * <p>An empty set does not imply that the credentials are fully understood.
+   * There may be optional configuration that can be ignored if not understood.
+   *
+   * <p>Since {@code Feature} is an {@code enum}, {@code understoodFeatures}
+   * should generally be an {@link EnumSet}. {@code understoodFeatures} will not
+   * be modified.
+   */
+  public Set<Feature> incomprehensible(Set<Feature> understoodFeatures) {...}
+
+  /* Other methods can be added to support various Features */
+
+  public enum Feature {}
+
+  public static final class Builder {
+    public ChannelCredentials build() {...}
+  }
+}
+```
+
+An example `Feature` could be `CLIENT_CERTIFICATE`. When the `Feature` is
+added, methods `getCertificateChain()`, `getPrivateKey()`, `getPassword()` can
+be added. Observing the contents of those methods would be a requirement for
+`CLIENT_CERTIFICATE`. An implementation understanding `CLIENT_CERTIFICATE`
+might not support encrypted private keys and so could consider the credential
+unsupported if `getPassword()` returned non-`null`.
+
+To support migration from existing builder options, Netty and OkHttp will
+provide transport-specific credentials. Note that neither transport will support
+reading the configuration after creation, except by the transport.
+
+```java
+package io.grpc.netty;
+
+/** A credential with full control over the security handshake. */
+public final class NettySslContextChannelCredentials {
+  private NettySslContextChannelCredentials() {}
+
+  /**
+   * Create a credential using Netty's SslContext as configuration. It must have
+   * been configured with {@link GrpcSslContexts}, but options could have been
+   * overridden.
+   */
+  public static ChannelCredentials create(SslContext sslContext) {...}
+}
+
+/** An insecure credential that upgrades from HTTP/1 to HTTP/2. */
+public final class InsecureFromHttp1ChannelCredentials {
+  private InsecureFromHttp1ChannelCredentials() {}
+
+  /** Creates an insecure credential that will upgrade from HTTP/1 to HTTP/2. */
+  public static ChannelCredentials create() {...}
+}
+```
+
+```java
+package io.grpc.okhttp;
+
+/** A credential with full control over the SSLSocketFactory. */
+public final class SslSocketFactoryChannelCredentials {
+  private SslSocketFactoryChannelCredentials() {}
+
+  public static ChannelCredentials create(SSLSocketFactory factory) {...}
+}
+```
+
+To support very advanced use cases, Netty will provide a credential that wraps
+a `ProtocolNegotiator`. This allows implementations like ALTS and XDS to use
+internal APIs without forcing their users to use experimental or internal APIs,
+as their users would just interact with `ChannelCredentials`.
+
+```java
+package io.grpc.netty;
+
+final class NettyChannelCredentials extends ChannelCredentials {
+  private NettyChannelCredentials(ProtocolNegotiator.Factory negotiator) {...}
```

Comment:
nit: I haven't found any `ProtocolNegotiator.Factory` but `NettyChannelBuilder.ProtocolNegotiatorFactory` unless it's a new class proposed.

Reply:
The existing class isn't quite right for this usage, simply because the old class is on the builder. So yes, it'd be some new class, but it would be similar to NettyChannelBuilder.ProtocolNegotiatorFactory. That's what was done in the implementation PR

Reply:
SG